multimodal/src/autogluon/multimodal/learners/object_detection.py

```diff
@@ -0,0 +1,733 @@
+        if is_train:
+            val_use_training_mode = (self._problem_type == OBJECT_DETECTION) and (self._validation_metric_name != MAP)
+            datamodule_kwargs.update(
+                dict(validate_data=self._tuning_data, val_use_training_mode=val_use_training_mode)
+            )
+            if (
+                self._problem_type == OBJECT_DETECTION
+                and model_config is not None
+                and MULTI_IMAGE_MIX_DATASET in model_config
+            ):
+                train_dataset = MultiImageMixDataset(
+                    data=self._train_data,
+                    preprocessor=[df_preprocessor],
+                    processors=[data_processors],
+                    model_config=model_config,
+                    id_mappings=None,
+                    is_training=True,
+                )
+                datamodule_kwargs.update(dict(train_dataset=train_dataset))
+            else:
+                datamodule_kwargs.update(dict(train_data=self._train_data))
+        else:
+            datamodule_kwargs.update(dict(predict_data=predict_data))
+
+        datamodule = BaseDataModule(**datamodule_kwargs)
+        return datamodule
+
+    def get_strategy_per_run(self, num_gpus, config):
+        if num_gpus <= 1:
+            strategy = "auto"
+        else:
+            strategy = DDP
+
+        return strategy
+
+    def update_num_gpus_by_strategy(self, strategy, num_gpus):
+        if strategy == DDP and self._fit_called:
+            num_gpus = 1  # While using DDP, we can only use single gpu after fit is called
+
+        return num_gpus
+
+    def get_optimization_kwargs_per_run(self, config, validation_metric, custom_metric_func):
+        return dict(
+            optim_type=config.optimization.optim_type,
+            lr_choice=config.optimization.lr_choice,
+            lr_schedule=config.optimization.lr_schedule,
+            lr=config.optimization.learning_rate,
+            lr_decay=config.optimization.lr_decay,
+            end_lr=config.optimization.end_lr,
+            lr_mult=config.optimization.lr_mult,
+            weight_decay=config.optimization.weight_decay,
+            warmup_steps=config.optimization.warmup_steps,
+            track_grad_norm=OmegaConf.select(config, "optimization.track_grad_norm", default=-1),
+            validation_metric=validation_metric,
+            validation_metric_name=self._validation_metric_name,
+            custom_metric_func=custom_metric_func,
+        )
+
+    def build_task_per_run(
+        self,
+        model: Optional[nn.Module] = None,
+        optimization_kwargs: Optional[dict] = None,
+        is_train=True,
+    ):
+        if is_train:
+            return MMDetLitModule(
+                model=model,
+                **optimization_kwargs,
+            )
+        else:
+            return MMDetLitModule(model=self._model)
+
+    def get_model_per_run(self, model, config):
+        if model is None:
+            model = create_fusion_model(
```

Comment:
Why are we not updating self._model here

Reply:
As the name tells, this is only for per run. In the current design, `fit_per_run` doesn't update any class attributes. The attribute updates will happen after `fit_per_run` is called. The reason is that all HPO runs share the same predictor and may call `fit_per_run` in parallel. So, we need to avoid update class attributes inside `fit_per_run`.